include/nix/DataFrame.hpp

```diff
@@ -0,0 +1,142 @@
+
+#include <nix/base/EntityWithSources.hpp>
+#include <nix/base/IDataFrame.hpp>
+
+#include <nix/Hydra.hpp>
+
+#include <string>
+#include <vector>
+
+namespace nix {
+
+class NIXAPI DataFrame : public base::EntityWithSources<base::IDataFrame> {
+public:
+
+    DataFrame()
+        : EntityWithSources()
+        {}
+
+    DataFrame(const std::shared_ptr<base::IDataFrame> &df)
+        : EntityWithSources(df)
+        {}
+
+    DataFrame(std::shared_ptr<base::IDataFrame> &&ptr)
+        : EntityWithSources(std::move(ptr))
+        {}
+
+    ndsize_t rows() const {
+        return backend()->rows();
+    }
+
+    void rows(ndsize_t n) {
+        return backend()->rows(n);
+    }
+
+    virtual std::vector<unsigned> nameToCol(const std::vector<std::string> &names) const {
+        return backend()->nameToCol(names);
+    }
+
+    virtual std::vector<std::string> colToName(const std::vector<unsigned> &cols) const {
+        return backend()->colToName(cols);
+    }
+
+    std::vector<Column> columns() const {
+        return backend()->columns();
+    }
+
+    void writeRow(ndsize_t row, const std::vector<Variant> &v) {
+        return backend()->writeRow(row, v);
+    }
+
+    void writeCell(ndsize_t row, unsigned col, const Variant &v) {
+        return this->writeCells(row, {{col, v}});
+    }
+
+    void writeCells(ndsize_t row, const std::vector<Cell> &cells) {
+        return backend()->writeCells(row, cells);
+    }
+
+    Variant readCell(ndsize_t row, unsigned col) const {
+        const std::vector<std::string> &cols = colToName({col});
+        if (cols.size() < 1) {
+            return Variant{};
+        }
+        return this->readCell(row, cols[0]);
+    }
+
+    Variant readCell(ndsize_t row, const std::string &col) const {
+        std::vector<Variant> cells = backend()->readCells(row, {col});
+        if (cells.size() < 1) {
+            return Variant{};
+        }
+
+        return cells[0];
+    }
+
+    std::vector<Variant> readCells(ndsize_t row, const std::vector<std::string> &cols) const {
+        return backend()->readCells(row, cols);
+    }
+
+    std::vector<Variant> readRow(ndsize_t row) {
+        return backend()->readRow(row);
+    }
+
+    template<typename T>
+        void writeColumn(int col,
+                         ndsize_t offset,
+                         const std::vector<T> &vals,
+                         ndsize_t count = 0) {
+        const Hydra<const std::vector<T>> hydra(vals);
+
+        if (count == 0)
+            count = vals.size();
+        else if (count > vals.size())
+            throw OutOfBounds("Requested to write more data than available");
+
+        DataType dtype = hydra.element_data_type();
+        backend()->writeColumn(col, offset, count, dtype, (const char *) hydra.data());
+    }
+
+    template<typename T>
+        void readColumn(int col,
+                        ndsize_t offset,
+                        std::vector<T> &vals,
+                        ndsize_t count = 0) {
```

Comment:
API: This should include a `bool` allowing the vector to be resized to the requested `count` and a way to say "the whole column"

Reply:
what is the policy now? appending to the vector? maybe put the vector as second arguments and the others get defaults, count == 0 could indicate that the whole column should be read

Reply:
currently we use the size of the vector, unless you specify count, then we assume the vector is big enough. We should at least make sure the latter is true I guess ;)

Reply:
I have added some commits with a new pair of functions. Maybe have a look if you like them.